### PR TITLE
Grau 10

### PR DESCRIPTION
**Pull request by DanMaguireData**

### CODE REVIEW

File: .github/workflows/run_tests.yaml
Comment (DanMaguireData) [inline .github/workflows/run_tests.yaml:1]: Pipeline executes our tests on pushes to main and on PRs we can discuss whats best here for us

File: .gitignore [hunk 1]
```diff
@@ -158,3 +158,4 @@ cython_debug/
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
+junit/test-results.xml
```
Comment (DANLENEHAN) [inline .gitignore:161]: If we didn't add this would test results actually be committed to the repo?

File: .github/workflows/run_tests.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        with:
+          python-version: '3.11'
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -e .
+          pip install -r requirements.txt
+      - name: Test with pytest
+        run: |
+          pip install -r tests/requirements.txt
```
Comment (DANLENEHAN) [inline .github/workflows/run_tests.yaml:26]: Can this one be in the 'Install dependencies'

File: alembic.ini
Comment (DANLENEHAN) [inline alembic.ini:1]: All these changes were probably a pre-commit deletion? I'd say we revert this and leave them in can we ignore the alembic folder?

File: development/api_interaction.ipynb [hunk 3]
```diff
@@ -181,7 +181,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.11.3"
+   "version": "3.11.4"
```
Comment (DANLENEHAN) [inline development/api_interaction.ipynb:184]: spooky

File: grau/blueprints/user/routes.py [hunk 4]
```diff
@@ -3,7 +3,6 @@
 from flask_sqlalchemy_session import current_session
 
 from grau.blueprints.user import functions
-from grau.blueprints.user.functions import attempt_login, attempt_logout
```
Comment (DANLENEHAN) [inline grau/blueprints/user/routes.py:6]: Why remove the explicit imports and references

File: grau/blueprints/user_stats/functions.py [hunk 5]
```diff
@@ -75,13 +73,15 @@ def update_user_stat(
     user_stats = get_user_stat(
         db_session, user_stats_dict["user_id"], user_stats_dict["id"]
     )
-    if not user_stats:
+    if user_stats is None:
```
Comment (DANLENEHAN) [inline grau/blueprints/user_stats/functions.py:76]: can just do if user_stats?

File: pyproject.toml [hunk 6]
```diff
@@ -6,7 +6,7 @@ dependencies = [
 		"Flask==2.0.1",
 		"Flask-SQLAlchemy-Session==1.1",
 		"pydantic==2.0",
-		"pydantic-extra-types-2.0.0",
```
Comment (DANLENEHAN) [inline pyproject.toml:9]: :(

File: tests/conftest.py [hunk 7]
```diff
@@ -86,21 +93,69 @@ def db_session(session_factory):
...
+            # Set user_id to 1 if not provided
+            if user_id is None:
+                user_id = 1
+            user_stat = user_stats_factory(user_id=user_id)
+
         _, response_code = create_user_stats(db_session, user_stat)
         if response_code != 201:
             raise Exception(f"Failed to insert user: {user_stat}")
+        # Get last user stat added to the db
+        # This is done bto retrive the id that was created for
```
Comment (DANLENEHAN) [inline tests/conftest.py:151]: 'bto' -> 'to'

File: tests/test_grau/blueprints/test_blueprint_integration.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+                "password": test_user["password"],
+            },
+            expected_response=b"Login successful"
+            if successful
+            else b"Login failed, invalid credentials",
+            expected_response_code=200 if successful else 400,
+            method="POST",
+        )
+
+    def test_user_session_management_happy_path(
```
Comment (DANLENEHAN) [inline tests/test_grau/blueprints/test_blueprint_integration.py:114]: Very clean

File: tests/test_grau/blueprints/test_blueprint_integration.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
+"""
+Integration scenarios testing user login, logout
+and session handling capabilities
+"""
+from typing import Dict, Union
+
+
+class IntegrationTestStep:
```
Comment (DANLENEHAN) [inline tests/test_grau/blueprints/test_blueprint_integration.py:8]: Is this expected to change from one module to another

File: tests/test_grau/blueprints/.DS_Store
Comment (DANLENEHAN) [inline tests/test_grau/blueprints/.DS_Store:1]: Is this needed?

File: tests/test_grau/blueprints/user/test_functions.py [hunk 10]
```diff
@@ -1,57 +1,131 @@
...
+        """
+        Tests get_user function
+        """
+        # Given
+        test_user = insert_user()
+
+        # when
+        result = functions.get_user(db_session, email=test_user.email)
+        # Then
+        db_class_comparison(result, test_user)
```
Comment (DANLENEHAN) [inline tests/test_grau/blueprints/user/test_functions.py:28]: like

File: requirements.txt [hunk 11]
```diff
@@ -5,7 +5,6 @@ configparser==5.3.0
 Flask==2.0.1
 Flask-Cors==3.0.10
 Flask-SQLAlchemy-Session==1.1
-freezegun==1.2.2
```
Comment (DANLENEHAN) [inline requirements.txt:8]: Why are we removing freezgun? Looks like we're using it in confest?

**Review by DANLENEHAN — COMMENTED** [review]
Minor comments, mainly for my understanding. Very clean testing solutions :()

**Review by DANLENEHAN — APPROVED** [review]

**Pull request merged**